abstract dataclass inheritance gives `Only concrete class can be given where "Type[Abstract]" is expected`

mypy 0.620, `--strict`.

Sample code:

```python3
import abc
from dataclasses import dataclass


@dataclass  # error: Only concrete class can be given where "Type[Abstract]" is expected
class Abstract(metaclass=abc.ABCMeta):
    a: str

    @abc.abstractmethod
    def c(self) -> None:
        pass


@dataclass
class Concrete(Abstract):
    b: str

    def c(self) -> None:
        pass


instance = Concrete(a='hello', b='world')
```

Note that this error *only* occurs if at least one abstract method is defined on the abstract class. Removing the abstract method `c` (or making it non-abstract) makes the error go away.

@JelleZijlstra 

My production code cut to be smaller:
```python
import dataclasses as dc
from abc import ABCMeta, abstractmethod


@dc.dataclass(frozen=True)
class CanaryConfiguration(metaclass=ABCMeta):
    @abstractmethod
    def is_fast_track(self) -> bool:
        """
        """


@dc.dataclass(frozen=True)
class CanaryConfigurationEcs(CanaryConfiguration):

    def is_fast_track(self) -> bool:
        return True


@dc.dataclass(frozen=True)
class Canary:
    configuration: CanaryConfiguration


def test_me() -> bool:
    c: Canary = Canary(CanaryConfigurationEcs())
    return c.configuration.is_fast_track()
```

Console output: 
```bash
→ mypy --version
mypy 0.800
→ mypy test.py
test.py:5: error: Only concrete class can be given where "Type[CanaryConfiguration]" is expected
Found 1 error in 1 file (checked 1 source file)
```

In actual code it additionally complains that the `return` statement in `test_me()` returns `Any` instead of  `bool` but somehow if I cut the business logic I can no longer reproduce it.

Adding one more (hopefully) valid use case

```python
from dataclasses import dataclass, fields, is_dataclass
from typing import Protocol, Optional, cast

@dataclass
class MyConfig:
    v1: str
    v2: 'Optional[str]' = None

@dataclass
class GlobalConfig:
    MyConfig: MyConfig

class SourceList(Protocol):
    def get(self, key: str) -> str:
        pass

class ConfigProtocol(Protocol):  # this is one option, but just being able to type check dataclass would be fine
    __dataclass_fields__: dict

class Configuration:
    def __init__(
        self,
        source_list: SourceList,
        config_class: ConfigProtocol = GlobalConfig
    ):
        self._config_sources = source_list

        for f in fields(self._config_class):
            if is_dataclass(f.type):
                setattr(
                    self,
                    f.name,
                    Configuration(
                        source_list=source_list,
                        config_class=cast(ConfigProtocol, f.type)
                    )
                )

    def __getattr__(self, name):
        if value := self._config_sources.get(name):
            return value
        raise AttributeError()
```
mypy output:
```
> mypy --version
mypy 0.800
> mypy test.py 
test.py:24: error: Incompatible default for argument "config_class" (default has type "Type[GlobalConfig]", argument has type "ConfigProtocol")
Found 1 error in 1 file (checked 1 source file)
```

Define dataclasses for config objects in modules and sub-modules, which can be instantiated and used directly, or as shown here, used as templates to build a more complex config class.

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -409,6 +409,11 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         # of giving a note on possibly missing "await". It is used to avoid infinite recursion.
         self.checking_missing_await = False
 
+        # While this is True, allow passing an abstract class where Type[T] is expected.
+        # although this is technically unsafe, this is desirable in some context, for
+        # example when type-checking class decorators.
+        self.allow_abstract_call = False
+
     @property
     def type_context(self) -> List[Optional[Type]]:
         return self.expr_checker.type_context
@@ -2027,9 +2032,12 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
                     # TODO: Figure out how to have clearer error messages.
                     # (e.g. "class decorator must be a function that accepts a type."
+                    old_allow_abstract_call = self.allow_abstract_call
+                    self.allow_abstract_call = True
                     sig, _ = self.expr_checker.check_call(
                         dec, [temp], [nodes.ARG_POS], defn, callable_name=fullname
                     )
+                    self.allow_abstract_call = old_allow_abstract_call
                 # TODO: Apply the sig to the actual TypeInfo so we can handle decorators
                 # that completely swap out the type.  (e.g. Callable[[Type[A]], Type[B]])
         if typ.is_protocol and typ.defn.type_vars:

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -1914,6 +1914,7 @@ class ExpressionChecker(ExpressionVisitor[Type]):
             and (caller_type.type_object().is_abstract or caller_type.type_object().is_protocol)
             and isinstance(callee_type.item, Instance)
             and (callee_type.item.type.is_abstract or callee_type.item.type.is_protocol)
+            and not self.chk.allow_abstract_call
         ):
             self.msg.concrete_only_call(callee_type, context)
         elif not is_subtype(caller_type, callee_type, options=self.chk.options):

--- a/test-data/unit/check-abstract.test
+++ b/test-data/unit/check-abstract.test
@@ -1018,3 +1018,15 @@ d = my_abstract_types['B']()  # E: Cannot instantiate abstract class "MyAbstract
 d.do()
 
 [builtins fixtures/dict.pyi]
+
+[case testAbstractClassesWorkWithGenericDecorators]
+from abc import abstractmethod, ABCMeta
+from typing import Type, TypeVar
+
+T = TypeVar("T")
+def deco(cls: Type[T]) -> Type[T]: ...
+
+@deco
+class A(metaclass=ABCMeta):
+    @abstractmethod
+    def foo(self, x: int) -> None: ...

